File: src/desktop/WordsNote.Desktop/Services/SyncSnapshot.cs
using WordsNote.Desktop.Models;

namespace WordsNote.Desktop.Services;

public static class SyncSnapshot
{
    public static string CreateDeckMatchKey(string title)
    {
        return Normalize(title);
    }

    public static string CreateDeckFingerprint(StudyDeck deck)
    {
        return string.Join("::", CreateDeckMatchKey(deck.Title), Normalize(deck.Description));
    }

    public static string CreateCardMatchKey(string collectionId, string front, string back)
    {
        return string.Join("::", Normalize(collectionId), Normalize(front), Normalize(back));
    }

    public static string CreateCardFingerprint(StudyCard card, string? collectionIdOverride = null)
    {
        var collectionId = string.IsNullOrWhiteSpace(collectionIdOverride)
            ? card.CollectionId
            : collectionIdOverride;

        return string.Join(
            "::",
            CreateCardMatchKey(collectionId, card.Front, card.Back),
            Normalize(card.Hint),
            NormalizeTags(card.Tags));
    }

    private static string Normalize(string? value)
    {
        return (value ?? string.Empty).Trim().ToLowerInvariant();
    }

    private static string NormalizeTags(IEnumerable<string>? tags)
    {
        if (tags is null)
        {
            return string.Empty;
        }

        return string.Join(
            "|",
            tags.Select(Normalize)
                .Where(tag => !string.IsNullOrWhiteSpace(tag))
                .OrderBy(tag => tag, StringComparer.Ordinal));
    }
}

File: src/extension/src/services/syncSnapshot.js
function normalizeSyncText(value) {
  return String(value || '').trim().toLowerCase();
}

function normalizeTags(tags) {
  return [...(Array.isArray(tags) ? tags : [])]
    .map((tag) => normalizeSyncText(tag))
    .filter(Boolean)
    .sort()
    .join('|');
}

export function createDeckMatchKey(title) {
  return normalizeSyncText(title);
}

export function createDeckFingerprint(deck) {
  return [createDeckMatchKey(deck.title), normalizeSyncText(deck.description)].join('::');
}

export function createCardMatchKey(collectionId, front, back) {
  return [
    normalizeSyncText(collectionId),
    normalizeSyncText(front),
    normalizeSyncText(back),
  ].join('::');
}

export function createCardFingerprint(card) {
  return [
    createCardMatchKey(card.collectionId, card.front, card.back),
    normalizeSyncText(card.hint || ''),
    normalizeTags(card.tags),
  ].join('::');
}
